index.d.ts

```diff
@@ -0,0 +1,23 @@
+export type LazyTrackedArgs = {
+  positional?: Array<unknown>;
+  named?: Record<string, unknown>;
+};
+
+type ConstructorFn<Args extends LazyTrackedArgs> = (() => Args) | (() => Args['positional']);
+
+export const use: PropertyDecorator;
+export class Resource<Args extends LazyTrackedArgs> {
+  protected args: Args;
+
+  // This is a lie, but makes the call site nice.
+  // Resources should not define a constructor.
+  constructor(fn: ConstructorFn<Args>);
+
+  get value(): unknown;
```

Comment:
Because of how types don't work with decorators, I've added a helper function for getting the type of the value

Reply:
Yeah, but one could be using the resource without the decorator. But anyway, I think if the resource has the types, this will be overwritten by the downstream implementation.

Reply:
>  but one could be using the resource without the decorator.

huh, I didn't know people would want to do that.

>  But anyway, I think if the resource has the types, this will be overwritten by the downstream implementation.

confirmed in playground demo.

Reply:
I just saw in the tests that `value` _doesn't_ need to be a getter, and you could _just set it in the lifecycle methods_.

_interesting_?

Reply:
value is now abstract readonly, like in your sample.

TIL, that `abstract readonly` works for both getters and properties :D